orchestrator/test_runner/src/erc_721_happy_path.rs

```diff
@@ -0,0 +1,176 @@
+use crate::utils::*;
+use crate::MINER_ADDRESS;
+use crate::MINER_PRIVATE_KEY;
+use crate::OPERATION_TIMEOUT;
+use crate::TOTAL_TIMEOUT;
+use clarity::{Address as EthAddress, Uint256};
+use deep_space::address::Address as CosmosAddress;
+use deep_space::Contact;
+use ethereum_gravity::send_erc721_to_cosmos::send_erc721_to_cosmos;
+use ethereum_gravity::utils::get_gravity_sol_address;
+use ethereum_gravity::utils::get_valset_nonce;
+use gravity_utils::error::GravityError;
+use std::time::Duration;
+use std::time::Instant;
+use web30::client::Web3;
+
+pub async fn erc721_happy_path_test(
+    web30: &Web3,
+    contact: &Contact,
+    keys: Vec<ValidatorKeys>,
+    gravity_address: EthAddress,
+    gravity_erc721_address: EthAddress,
+    erc721_address: EthAddress,
+    validator_out: bool,
+) {
+    let grav_sol_address_in_erc721 =
+        get_gravity_sol_address(gravity_erc721_address, *MINER_ADDRESS, web30)
+            .await
+            .unwrap();
+
+    info!("ERC721 address is {}", erc721_address);
+    info!("Miner address is {}", *MINER_ADDRESS);
+    info!("Miner pk is {}", *MINER_PRIVATE_KEY);
+    info!(
+        "grav_sol_address_in_erc721 is {}",
+        grav_sol_address_in_erc721
+    );
+    info!("Gravity address is {}", gravity_address);
+    info!("GravityERC721 address is {}", gravity_erc721_address);
+
+    assert_eq!(grav_sol_address_in_erc721, gravity_address);
+
+    let no_relay_market_config = create_default_test_config();
+    start_orchestrators(
+        keys.clone(),
+        gravity_address,
+        validator_out,
+        no_relay_market_config,
+    )
+    .await;
+
+    // generate an address for coin sending tests, this ensures test imdepotency
+    let user_keys = get_user_key();
```

Comment:
you could transfer the erc721 to the eth address here from the miner address. This would let you test your erc721 transfer function. Then you send it from the user eth address to gravityERC721.

Reply:
I added this function to test a transfer:
```
pub async fn test_erc721_transfer_utils(
    web30: &Web3,
    gravity_erc721_address: EthAddress,
    erc721_address: EthAddress,
    token_id: Uint256
)  {
    match web30
    .approve_erc721_transfers(
        erc721_address,
        *MINER_PRIVATE_KEY,
        gravity_erc721_address,
        token_id.clone(),
        None,
        vec![]
    )
    .await 
    {
        Ok(_) => {
            info!("Successfully called ERC721 transfer util for {}!", token_id)
        }
        Err(_) => {
            panic!("Failed ERC721 transfer util for {}!", token_id)
        }
    }
}
```